lib/common/socket.c

```diff
@@ -1507,6 +1507,29 @@ void h2o_ssl_register_npn_protocols(SSL_CTX *ctx, const char *protocols)
 
 #endif
 
+int h2o_ssl_peer_verify_callback (int preok, X509_STORE_CTX *store_ctx)
+{
+    char data[256];
+
+    /* TODO: How much details are verified by default by openssl? */
+    /* if preverify fails get more details */
+    if (!preok)
+    {
+        X509 *cert = X509_STORE_CTX_get_current_cert(store_ctx);
+        int  depth = X509_STORE_CTX_get_error_depth(store_ctx);
+        int  err = X509_STORE_CTX_get_error(store_ctx);
+
+        fprintf(stderr, "-Error with certificate at depth: %i\n", depth);
+        X509_NAME_oneline(X509_get_issuer_name(cert), data, 256);
+        fprintf(stderr, "  issuer   = %s\n", data);
+        X509_NAME_oneline(X509_get_subject_name(cert), data, 256);
+        fprintf(stderr, "  subject  = %s\n", data);
+        fprintf(stderr, "  err %i:%s\n", err, X509_verify_cert_error_string(err) );
+    }
+
+    return preok;
+}
+
```

Comment:
I can see the intent of logging details to figure out the cause of authentication failure, I am not sure if we want to bake that into our library-level API.

To paraphrase, I think that this sort of callback that logs diagnosing information should be defined in src/main.c. Then, we could add a config knob to turn on / off this logging (as well as specifying to where the log should be written, if such knob is considered necessary).

Reply:
I agree that the logging should be controlled, do we have a debug build flag so that logging only enabled in debug build?

Reply:
If debugging is the purpose of logging these information, the answer for us would be to create USDT probe points, because they would enable us to attach probes while the server is running in production to figure out what is going on. picotls-probes.d and PTLS_PROBE macro would be good places to look at (if the change is going into pictols; there are similar files / macros in h2o as well).

Reply:
For current use case no additional debugging info is needed, removed all logging.

Reply:
I'll remove the callback. Will revisit or bring it back if customized verification is required.